src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs

```diff
@@ -2,35 +2,96 @@
         where TCachedMessage : struct
     {
+        protected readonly int defaultMaxAddCount;
+        protected readonly PooledQueueCache<EventData, TCachedMessage> cache;
+        private readonly CachePressureMonitor cachePressureMonitor;
+
         protected IStreamQueueCheckpointer<string> Checkpointer { private set; get; }
 
-        protected EventHubQueueCache(IStreamQueueCheckpointer<string> checkpointer, ICacheDataAdapter<EventData, TCachedMessage> cacheDataAdapter, ICacheDataComparer<TCachedMessage> comparer)
-            : base(cacheDataAdapter, comparer)
+        protected EventHubQueueCache(int defaultMaxAddCount, IStreamQueueCheckpointer<string> checkpointer, ICacheDataAdapter<EventData, TCachedMessage> cacheDataAdapter, ICacheDataComparer<TCachedMessage> comparer)
         {
-            cacheDataAdapter.PurgeAction = Purge;
+            this.defaultMaxAddCount = defaultMaxAddCount;
             Checkpointer = checkpointer;
-            OnPurged = CheckpointOnPurged;
+
+            cache = new PooledQueueCache<EventData, TCachedMessage>(cacheDataAdapter, comparer);
+            cacheDataAdapter.PurgeAction = cache.Purge;
+            cache.OnPurged = CheckpointOnPurged;
+
+            cachePressureMonitor = new CachePressureMonitor();
         }
 
+        protected abstract string GetOffset(TCachedMessage lastItemPurged);
+
+        // cachePressureContribution should be a double between 0-1, indicating how much danger the item is of being removed from the cache.  0 indicating  no danger, 1 indecating removal is imminent.
+        protected abstract bool TryCalculateCachePressureContribution(StreamSequenceToken token, out double cachePressureContribution);
+
         public void Dispose()
         {
-            OnPurged = null;
+            cache.OnPurged = null;
         }
 
-        protected abstract string GetOffset(TCachedMessage lastItemPurged);
-
         private void CheckpointOnPurged(TCachedMessage lastItemPurged)
         {
             Checkpointer.Update(GetOffset(lastItemPurged), DateTime.UtcNow);
         }
+
+        public int MaxAddCount { get { return cachePressureMonitor.IsUnderPressure() ? 0 : defaultMaxAddCount; } }
+
+        public void Add(EventData message)
+        {
+            cache.Add(message);
+        }
+
+        public object GetCursor(IStreamIdentity streamIdentity, StreamSequenceToken sequenceToken)
+        {
+            return cache.GetCursor(streamIdentity, sequenceToken);
+        }
+
+        public bool TryGetNextMessage(object cursorObj, out IBatchContainer message)
+        {
+            if (!cache.TryGetNextMessage(cursorObj, out message))
+                return false;
+            double cachePressureContribution;
+            if (TryCalculateCachePressureContribution(message.SequenceToken, out cachePressureContribution))
+            {
+                cachePressureMonitor.RecordCachePressureContribution(cachePressureContribution);
```

Comment:
Your understanding of this is correct.  In the more detailed description I provided, at Gabriel's request, I included two streams, one which read faster and contributed more to the pressure.  I included this example to help illuminate such peculiarities.

This approach was selected because it's fast and simple.  Testing is needed to determine if it's a viable algorithm.

Most alternative approaches I've explored come with a cost.  For instance, tracking each cursor's most recent contribution incurs the cost of a heavier data structure (dictionary), a lookup per contribution, a more expensive aggregation, and a more expensive reset after each aggregated measurement.  While these costs probably wouldn't effect performance noticeably, they are overhead that shouldn't be incurred unless we've reason to believe lighter algorithms are not sufficient.

I'm very open to alternatives if there is reason to believe the value of the measurement relative to the cost of the operation is better.  In my view this is a very interesting space to explore, but given my current timelines, I will likely settle for the first fast simple algorithm that works reasonably well.